### PR TITLE
Feature/dev deployable

### PR DESCRIPTION
**Pull request by nk-coding**
- adds jwt (actually jws) support to public api
- removes configuration from this repository
- adds documentation about the modules

### CODE REVIEW

File: api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt [hunk 1]
```diff
@@ -1,34 +1,51 @@
...
+
     override suspend fun generateContextMap(request: ServerRequest): Map<*, Any> {
-        //TODO use authentication as soon as available
-        val userId = request.headers().firstHeader("Authorization")
-        val additionalContextEntries = if (userId == null) {
-            emptyMap()
+        val token = request.headers().firstHeader("Authorization")
+        val additionalContextEntries = if (token == null) {
+            if (gropiusPublicApiConfigurationProperties.debugNoAuthentication) {
+                emptyMap()
```
Comment (ModellbahnFreak) [inline api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt:42]: Shouldn't debug mode add admin for every request?

File: api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt [hunk 2]
```diff
@@ -39,4 +56,22 @@ class GropiusGraphQLContextFactory(
...
 
+    /**
+     * Verifies a jwt and returns the [GropiusUser] defined by the id in subject
+     *
+     * @param token the jwt, possibly starting with `"Bearer "`
+     * @return the user defined by the id in subject
+     * @throws IllegalStateException if the user does not exist or if the jwt is invalid
+     */
+    private suspend fun verifyToken(token: String): GropiusUser {
+        val tokenWithoutBearer = token.replace("Bearer ", "")
```
Comment (ModellbahnFreak) [inline api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt:67]: Must be case insensitive matching for the auth scheme, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization#directives

Comment (ModellbahnFreak) [reply api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt:67]: Maybe as idea in future for easier integation with a frontend, we could think about supporting tokens in cookie header.

**Review by ModellbahnFreak — CHANGES_REQUESTED** [review]

File: api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt (hunk 2 above)
Comment (nk-coding) [reply api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt]: 69d97778ef86e49b34a43e423b0f290a463c5073

File: api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt (hunk 1 above)
Comment (nk-coding) [reply api-public/src/main/kotlin/gropius/graphql/GropiusGraphQLContextFactory.kt:42]: efac170f5511a19759d36954aa57d29c2ce9e6a2  
I forgot the necessary code a few lines below  
Regarding your question, it depends. Without a token, no authorization context is put into the map. By the query part, this is interpreted as no authorization required, therefore allowing all READ access to everything.
Mutations however require a token, as mutations need to know who performs the mutation (at least in a lot of cases).
The fix mentioned by the commit above fixes this behavior, as now debugNoAuthentication behaves like isAdmin for the authenticated user

**Review by chriku — APPROVED** [review]

**Review by ModellbahnFreak — APPROVED** [review]

**Pull request merged**